File: repo-client/src/main/java/io/lionweb/repoclient/LowLevelRepoClient.java
package io.lionweb.repoclient;

import com.google.gson.*;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.*;
import java.util.concurrent.TimeUnit;
import okhttp3.*;
import okio.Buffer;
import okio.BufferedSink;
import okio.GzipSink;
import okio.Okio;

public class LowLevelRepoClient {

  private final String hostname;
  private final int port;
  private final String authorizationToken;
  private String clientID;
  private String repository;
  private final long connectTimeoutInSeconds;
  private final long callTimeoutInSeconds;
  private final boolean debug;

  private final OkHttpClient httpClient;

  private static final MediaType JSON = MediaType.get("application/json");
  private static final MediaType PROTOBUF = MediaType.get("application/protobuf");
  private static final MediaType FLATBUFFERS = MediaType.get("application/flatbuffers");

  public LowLevelRepoClient() {
    this("localhost", 3005, null, "GenericKotlinBasedLionWebClient", "default", 60, 60, false);
  }

  public LowLevelRepoClient(
      String hostname,
      int port,
      String authorizationToken,
      String clientID,
      String repository,
      long connectTimeoutInSeconds,
      long callTimeoutInSeconds,
      boolean debug) {
    this.hostname = hostname;
    this.port = port;
    this.authorizationToken = authorizationToken;
    this.clientID = clientID;
    this.repository = repository;
    this.connectTimeoutInSeconds = connectTimeoutInSeconds;
    this.callTimeoutInSeconds = callTimeoutInSeconds;
    this.debug = debug;

    this.httpClient =
        new OkHttpClient.Builder()
            .connectTimeout(connectTimeoutInSeconds, TimeUnit.SECONDS)
            .readTimeout(callTimeoutInSeconds, TimeUnit.SECONDS)
            .writeTimeout(callTimeoutInSeconds, TimeUnit.SECONDS)
            .callTimeout(callTimeoutInSeconds, TimeUnit.SECONDS)
            .build();
  }

  public void createRepository(boolean history) throws IOException {
    String url = String.format("http://%s:%d/createRepository?history=%s", hostname, port, history);
    Request.Builder rq = new Request.Builder().url(addClientIdQueryParam(url));
    rq = considerAuthenticationToken(rq);
    Request request = rq.post(RequestBody.create(new byte[0], null)).build();

    try (Response response = httpClient.newCall(request).execute()) {
      if (response.code() != HttpURLConnection.HTTP_OK) {
        throw new RuntimeException(
            "DB initialization failed, HTTP " + response.code() + ": " + response.body().string());
      }
    }
  }

  public void deletePartition(String nodeID) throws IOException {
    String bodyJson = "[\"" + nodeID + "\"]";
    RequestBody body = RequestBody.create(bodyJson, JSON);
    Request.Builder rq =
        new Request.Builder()
            .url(
                addClientIdQueryParam(
                    "http://" + hostname + ":" + port + "/bulk/deletePartitions"));
    rq = considerAuthenticationToken(rq);
    Request request = rq.post(body).build();

    try (Response response = httpClient.newCall(request).execute()) {
      if (response.code() != HttpURLConnection.HTTP_OK) {
        if (debug) {
          System.out.println("  Response: " + response.code());
          System.out.println("  Response: " + response.body().string());
        }
        throw new RuntimeException(
            "Request failed with code " + response.code() + ": " + response.body().string());
      }
    }
  }

  public String getPartitionIDs() throws IOException {
    String url = "http://" + hostname + ":" + port + "/bulk/listPartitions";
    Request.Builder rq =
        new Request.Builder().url(addRepositoryQueryParam(addClientIdQueryParam(url)));
    rq = considerAuthenticationToken(rq);
    Request request =
        rq.addHeader("Accept-Encoding", "gzip").post(RequestBody.create(new byte[0], null)).build();

    try (Response response = httpClient.newCall(request).execute()) {
      if (response.code() == HttpURLConnection.HTTP_OK) {
        return Objects.requireNonNull(response.body()).string();
      } else {
        throw new RuntimeException("Got back " + response.code() + ": " + response.body().string());
      }
    }
  }

  public String retrieve(List<String> rootIds, int limit) throws IOException {
    if (rootIds.isEmpty() || rootIds.stream().anyMatch(String::isEmpty)) {
      throw new IllegalArgumentException("Root IDs must not be empty or blank");
    }

    String bodyJson =
        "{\"ids\":["
            + String.join(", ", rootIds.stream().map(id -> "\"" + id + "\"").toArray(String[]::new))
            + "]}";
    HttpUrl.Builder urlBuilder =
        HttpUrl.parse("http://" + hostname + ":" + port + "/bulk/retrieve").newBuilder();
    urlBuilder.addQueryParameter("depthLimit", String.valueOf(limit));
    urlBuilder.addQueryParameter("clientId", clientID);
    urlBuilder.addQueryParameter("repository", repository);

    Request.Builder rq = new Request.Builder().url(urlBuilder.build());
    rq = considerAuthenticationToken(rq);
    Request request = rq.post(RequestBody.create(bodyJson, JSON)).build();

    try (Response response = httpClient.newCall(request).execute()) {
      if (response.code() == HttpURLConnection.HTTP_OK) {
        return Objects.requireNonNull(response.body()).string();
      } else {
        throw new RuntimeException(
            "Error retrieving data: " + response.code() + ", " + response.body().string());
      }
    }
  }

  public Map<ClassifierKey, ClassifierResult> nodesByClassifier(Integer limit) throws IOException {
    HttpUrl.Builder urlBuilder =
        HttpUrl.parse("http://" + hostname + ":" + port + "/inspection/nodesByClassifier")
            .newBuilder();
    urlBuilder.addQueryParameter("clientId", clientID);
    if (limit != null) {
      urlBuilder.addQueryParameter("limit", String.valueOf(limit));
    }

    Request.Builder rq = new Request.Builder().url(urlBuilder.build());
    rq = considerAuthenticationToken(rq);
    Request request = rq.get().build();

    try (Response response = httpClient.newCall(request).execute()) {
      String body = Objects.requireNonNull(response.body()).string();
      if (response.code() != HttpURLConnection.HTTP_OK) {
        throw new RuntimeException("Request failed: " + response.code() + ": " + body);
      }

      JsonArray array = JsonParser.parseString(body).getAsJsonArray();
      Map<ClassifierKey, ClassifierResult> result = new HashMap<>();
      for (JsonElement element : array) {
        JsonObject obj = element.getAsJsonObject();
        ClassifierKey key =
            new ClassifierKey(
                obj.get("language").getAsString(), obj.get("classifier").getAsString());
        Set<String> ids = new HashSet<>();
        obj.get("ids").getAsJsonArray().forEach(idElem -> ids.add(idElem.getAsString()));
        int size = obj.get("size").getAsInt();
        result.put(key, new ClassifierResult(ids, size));
      }
      return result;
    }
  }

  // Remaining methods: nodeTree, bulkImportUsingJson, nodesStoringOperation, bulkImport etc.
  // ... (Let me know if you'd like the rest completed now)

  // ──────────────────────────────────────────────────────
  // 🔧 Helpers
  // ──────────────────────────────────────────────────────

  private HttpUrl addClientIdQueryParam(String rawUrl) {
    HttpUrl.Builder builder = HttpUrl.parse(rawUrl).newBuilder();
    builder.addQueryParameter("clientId", clientID);
    return builder.build();
  }

  private HttpUrl addRepositoryQueryParam(HttpUrl url) {
    return url.newBuilder().addQueryParameter("repository", repository).build();
  }

  private Request.Builder considerAuthenticationToken(Request.Builder builder) {
    return (authorizationToken == null)
        ? builder
        : builder.addHeader("Authorization", authorizationToken);
  }

  private RequestBody gzipCompress(RequestBody original) throws IOException {
    Buffer buffer = new Buffer();
    original.writeTo(buffer);

    RequestBody gzippedBody =
        new RequestBody() {
          @Override
          public MediaType contentType() {
            return original.contentType();
          }

          @Override
          public long contentLength() {
            return -1; // unknown
          }

          @Override
          public void writeTo(BufferedSink sink) throws IOException {
            GzipSink gzipSink = new GzipSink(sink);
            BufferedSink compressedSink = Okio.buffer(gzipSink);
            buffer.copyTo(compressedSink.buffer(), 0, buffer.size());
            compressedSink.close();
          }
        };

    return gzippedBody;
  }

  private Request.Builder addGZipCompressionHeader(Request.Builder builder) {
    return builder.addHeader("Content-Encoding", "gzip");
  }
}

File: repo-client/src/main/kotlin/io/lionweb/lioncore/kotlin/repoclient/LowLevelRepoClient.kt
package io.lionweb.lioncore.kotlin.repoclient

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.internal.EMPTY_REQUEST
import java.net.ConnectException
import java.net.HttpURLConnection
import java.util.concurrent.TimeUnit

/**
 * This is a low-level client for the LW Repo. It just does the requests, without any preparation or safe-guard.
 * Users of this library should use LionWebClient, which will delegate to this class.
 */
internal class LowLevelRepoClient(
    val hostname: String = "localhost",
    val port: Int = 3005,
    val authorizationToken: String? = null,
    var clientID: String = "GenericKotlinBasedLionWebClient",
    var repository: String = "default",
    val connectTimeOutInSeconds: Long = 60,
    val callTimeoutInSeconds: Long = 60,
    val debug: Boolean = false,
) {
    fun createRepository(history: Boolean = false) {
        val url = "http://$hostname:$port/createRepository?history=$history"
        val request: Request =
            Request.Builder()
                .url(url.addClientIdQueryParam())
                .considerAuthenticationToken()
                .post(EMPTY_REQUEST)
                .build()
        httpClient.newCall(request).execute().use { response ->
            if (response.code != HttpURLConnection.HTTP_OK) {
                throw RuntimeException("DB initialization failed, HTTP ${response.code}: ${response.body?.string()}")
            }
        }
    }

    fun deletePartition(nodeID: String) {
        val body: RequestBody = "[\"${nodeID}\"]".toRequestBody(JSON)
        val request: Request =
            Request.Builder()
                .url("http://$hostname:$port/bulk/deletePartitions".addClientIdQueryParam())
                .considerAuthenticationToken()
                .post(body)
                .build()
        httpClient.newCall(request).execute().use { response ->
            if (response.code != HttpURLConnection.HTTP_OK) {
                val body = response.body?.string()
                if (debug) {
                    println("  Response: ${response.code}")
                    println("  Response: $body")
                }
                throw RuntimeException("Request failed with code ${response.code}: $body")
            }
        }
    }

    fun getPartitionIDs(): String {
        val url = "http://$hostname:$port/bulk/listPartitions"
        val request: Request =
            Request.Builder()
                .url(url.addClientIdQueryParam())
                .considerAuthenticationToken()
                .addHeader("Accept-Encoding", "gzip")
                .post(EMPTY_REQUEST)
                .build()
        httpClient.newCall(request).execute().use { response ->
            if (response.code == HttpURLConnection.HTTP_OK) {
                val data =
                    (response.body ?: throw IllegalStateException("Response without body when querying $url")).string()
                return data
            } else {
                throw RuntimeException("Got back ${response.code}: ${response.body?.string()}")
            }
        }
    }

    fun retrieve(
        rootIds: List<String>,
        limit: Int,
    ): String {
        require(rootIds.isNotEmpty())
        require(rootIds.all { it.isNotBlank() })
        val body: RequestBody = "{\"ids\":[${rootIds.joinToString(", "){"\"$it\""}}] }".toRequestBody(JSON)
        val url = "http://$hostname:$port/bulk/retrieve"
        val urlBuilder = url.toHttpUrlOrNull()!!.newBuilder()
        urlBuilder.addQueryParameter("depthLimit", limit.toString())
        urlBuilder.addQueryParameter("clientId", clientID)
        urlBuilder.addQueryParameter("repository", repository)
        val request: Request =
            Request.Builder()
                .url(urlBuilder.build())
                .considerAuthenticationToken()
                .post(body)
                .build()
        httpClient.newCall(request).execute().use { response ->
            if (response.code == HttpURLConnection.HTTP_OK) {
                val data =
                    (response.body ?: throw IllegalStateException("Response without body when querying $url")).string()
                return data
            } else {
                throw RuntimeException(
                    "Something went wrong while querying $url: http code ${response.code}, body: ${response.body?.string()}",
                )
            }
        }
    }

    fun nodesByClassifier(limit: Int? = null): Map<ClassifierKey, ClassifierResult> {
        val url = "http://$hostname:$port/inspection/nodesByClassifier"
        val urlBuilder = url.toHttpUrlOrNull()!!.newBuilder()
        urlBuilder.addQueryParameter("clientId", clientID)
        if (limit != null) {
            urlBuilder.addQueryParameter("limit", limit.toString())
        }
        val request: Request =
            Request.Builder()
                .url(urlBuilder.build())
                .considerAuthenticationToken()
                .get()
                .build()
        httpClient.newCall(request).execute().use { response ->
            val body = response.body?.string()
            if (response.code != HttpURLConnection.HTTP_OK) {
                throw RuntimeException("DB initialization failed, HTTP ${response.code}: $body")
            }
            val data = JsonParser.parseString(body)
            val result = mutableMapOf<ClassifierKey, ClassifierResult>()
            data.asJsonArray.map { it.asJsonObject }.forEach { entry ->
                val classifierKey = ClassifierKey(entry["language"].asString, entry["classifier"].asString)
                val ids: Set<String> = entry["ids"].asJsonArray.map { it.asString }.toSet()
                result[classifierKey] = ClassifierResult(ids, entry["size"].asInt)
            }
            return result
        }
    }

    fun nodeTree(
        nodeIDs: List<String>,
        depthLimit: Int? = null,
    ): List<NodeInfo> {
        val url = "http://$hostname:$port/additional/getNodeTree"
        val urlBuilder = url.toHttpUrlOrNull()!!.newBuilder()
        if (depthLimit != null) {
            urlBuilder.addQueryParameter("depthLimit", depthLimit.toString())
        }
        urlBuilder.addQueryParameter("clientId", clientID)
        urlBuilder.addQueryParameter("repository", repository)
        val body = JsonObject()
        val ids = JsonArray()
        nodeIDs.forEach { ids.add(it) }
        body.add("ids", ids)
        val bodyJson = Gson().toJson(body)
        val builder =
            Request.Builder()
                .url(urlBuilder.build())
                .considerAuthenticationToken()
                .post(bodyJson.toRequestBody(JSON))

        val request: Request =
            builder
                .build()

        httpClient.newCall(request).execute().use { response ->
            val body = response.body?.string()
            if (response.code != HttpURLConnection.HTTP_OK) {
                throw RuntimeException("${response.code}: $body")
            }

            val data = JsonParser.parseString(body).asJsonObject.get("data").asJsonArray
            return data.map { it.asJsonObject }.map { dataElement ->
                val parent =
                    if (dataElement.has("parent") && dataElement.get("parent") !is JsonNull) {
                        dataElement.get(
                            "parent",
                        ).asString
                    } else {
                        null
                    }
                NodeInfo(dataElement.get("id").asString, parent, dataElement.get("depth").asInt)
            }
        }
    }

    fun bulkImportUsingJson(
        bodyJson: String,
        compress: Boolean = false,
    ) {
        val requestBody = bodyJson.toRequestBody(JSON).considerCompression(compress)
        bulkImport(requestBody, compress)
    }

    fun bulkImportUsingProtobuf(
        bytes: ByteArray,
        compress: Boolean = false,
    ) {
        val requestBody = bytes.toRequestBody(PROTOBUF).considerCompression(compress)
        bulkImport(requestBody, compress)
    }

    fun bulkImportUsingFlatBuffers(
        bytes: ByteArray,
        compress: Boolean = false,
    ) {
        val requestBody = bytes.toRequestBody(FLATBUFFERS).considerCompression(compress)
        bulkImport(requestBody, compress)
    }

    fun nodesStoringOperation(
        json: String,
        operation: String,
    ) {
        val body: RequestBody = json.compress()

        // TODO control with flag http or https
        val url = "http://$hostname:$port/bulk/$operation"
        val request: Request =
            Request.Builder()
                .url(url.addClientIdQueryParam().addRepositoryQueryParam())
                .considerAuthenticationToken()
                .addGZipCompressionHeader()
                .post(body)
                .build()
        try {
            httpClient.newCall(request).execute().use { response ->
                if (response.code != HttpURLConnection.HTTP_OK) {
                    val body = response.body?.string()
                    if (debug) {
                        println("  Response: ${response.code}")
                        println("  Response: $body")
                    }
                    throw RequestFailureException(url, json, response.code, body)
                }
            }
        } catch (e: ConnectException) {
            val jsonExcept =
                if (json.length > 10000) {
                    json.substring(0, 1000) + "..."
                } else {
                    json
                }
            throw RuntimeException("Cannot get answer from the client when contacting at URL $url. Body: $jsonExcept", e)
        }
    }

    private fun bulkImport(
        requestBody: RequestBody,
        compress: Boolean = false,
    ) {
        val url = "http://$hostname:$port/additional/bulkImport"
        val urlBuilder = url.toHttpUrlOrNull()!!.newBuilder()
        urlBuilder.addQueryParameter("clientId", clientID)
        urlBuilder.addQueryParameter("repository", repository)
        val builder =
            Request.Builder()
                .url(urlBuilder.build())
                .considerAuthenticationToken()
                .considerCompression(compress)
                .post(requestBody)

        val request: Request =
            builder
                .build()

        httpClient.newCall(request).execute().use { response ->
            val body = response.body?.string()
            if (response.code != HttpURLConnection.HTTP_OK) {
                throw RuntimeException("${response.code}: $body")
            }

            val success = JsonParser.parseString(body).asJsonObject.get("success").asBoolean
            if (!success) {
                throw RuntimeException("Request failed: $body")
            }
        }
    }

    private var httpClient: OkHttpClient =
        OkHttpClient.Builder()
            .callTimeout(
                callTimeoutInSeconds,
                TimeUnit.SECONDS,
            ).readTimeout(callTimeoutInSeconds, TimeUnit.SECONDS)
            .writeTimeout(callTimeoutInSeconds, TimeUnit.SECONDS)
            .connectTimeout(connectTimeOutInSeconds, TimeUnit.SECONDS).build()

    private fun String.addClientIdQueryParam(): HttpUrl {
        val urlBuilder = this.toHttpUrl().newBuilder()
        urlBuilder.addQueryParameter("clientId", clientID)
        return urlBuilder.build()
    }

    private fun Request.Builder.considerAuthenticationToken(): Request.Builder {
        return if (authorizationToken == null) {
            this
        } else {
            this.addHeader("Authorization", authorizationToken)
        }
    }

    private fun Request.Builder.considerCompression(compress: Boolean): Request.Builder {
        return if (compress) {
            this.addGZipCompressionHeader()
        } else {
            this
        }
    }

    private fun HttpUrl.addRepositoryQueryParam(): HttpUrl {
        val urlBuilder = this.newBuilder()
        urlBuilder.addQueryParameter("repository", repository)
        return urlBuilder.build()
    }
}

fun Request.Builder.addGZipCompressionHeader(): Request.Builder {
    this.addHeader("Content-Encoding", "gzip")
    return this
}

fun RequestBody.considerCompression(compress: Boolean): RequestBody {
    return if (compress) {
        this.compress()
    } else {
        this
    }
}
